python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,216 @@
+
+            annotated.append(
+                obj.options(
+                    scheduling_strategy=PlacementGroupSchedulingStrategy(
+                        placement_group=self.placement_group,
+                        # Max ensures that empty head bundles are correctly placed
+                        placement_group_bundle_index=max(0, i),
+                        placement_group_capture_child_tasks=True,
+                    ),
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    """Resource manager using placement groups as the resource backend.
+
+    Ray core does not emit events when resources are available. Instead, we
+    have to periodically request and update the scheduling state of the
+    placement groups.
+
+    Internally, the placement group lifecycle is like this:
+    - Resources are requested with ``request_resources()``
+    - A placement group is scheduled ("staged")
+    - A ``PlacementGroup.ready()`` is scheduled ("staging future")
+    - We update the scheduling state when we need to
+        (e.g. when ``has_resources_ready()`` is called)
+    - When staging futures resolved, a placement group is moved from "staging"
+        to "ready"
+    - When a resource request is canceled, we remove a placement group from "staging".
+        If there are not staged placement groups (because they are already "ready"),
+        we remove one from "ready" instead.
+    - When a resource is acquired, the pg is removed from "ready" and moved
+        to "acquired"
+    - When a resource is freed, the pg is removed from "acquired" and destroyed
+
+    Per default, placement group scheduling state is refreshed every time when
+    resource state is inquired, but not more often than once every ``update_interval_s``
+    seconds. Alternatively, staging futures can be retrieved (and awaited) with
+    ``get_resource_futures()`` and state update can be force with ``update_state()``.
+
+    Args:
+        update_interval_s: Minimum interval in seconds between updating scheduling
+            state of placement groups.
+
+    """
+
+    _resource_cls: AcquiredResource = PlacementGroupAcquiredResource
+
+    def __init__(self, update_interval_s: float = 0.1):
+        self._pg_to_request: Dict[PlacementGroup, ResourceRequest] = {}
+        self._request_to_staged_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+        self._request_to_ready_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+
+        self._staging_future_to_pg: Dict[ray.ObjectRef, PlacementGroup] = dict()
+        self._pg_to_staging_future: Dict[PlacementGroup, ray.ObjectRef] = dict()
+        self._acquired_pgs: Set[PlacementGroup] = set()
```

Comment:
do we really need all these separate maps?
I'd imagine something like
```
RequestAndPG = namedtuple("RequestAndPG", ["request", "pg", "state"])
self._pgs: List[RequestAndPG] = []
```
And you can update or filter by state easily all in this list.
It's preferable to keep the data structure simple and keep fewer states.
Easier to understand, and less error prone.
Reason I think we can use a simple list is that there will never be that many PGs to manage.
Even if you have hundreds of PGs, this should still be fast enough.

Reply:
Fwiw, the reason why we went with maps is that we actually observed scalability issues when running with 1000s of parallel trials. We scan through the states very often and efficient lookup did reduce the lag significantly.

The logic is not going to change anytime soon, so once implemented it should be easy to maintain. Since your review I've also updated the line comments and grouped the maps. 

If you're concerned about correctness, I can add a unit test to assert internal state is correct, if that helps?

Reply:
I'll add a unit test!